update full path and fix linux LIST empty

lawnvi opened the pull request:
Hello~

I compared an Android app that supports an FTP server and found that it sends an open data connection command before responding with the file list for the LIST command. Additionally, the Ubuntu client shows the files after I add them.

Changes in this PR:

1. Fixed the full path error, which did not consider file or directory names containing spaces or '-'.
2. Linux works as a client .
3. Implemented the UTF-8 option.

The Windows client still has some issues but basically satisfies my needs.
- Occasionally, the LIST command fails to respond because the dataSocket is null when using Windows as the server and Linux as the client, retrying may succeed.

lawnvi committed:
fix passive mode address

lawnvi committed:
fix full path, file names containing '-' or spaces, and windows path separators.

lawnvi committed:
impl opts utf8

abdelaziz-mahdy commented on lib/ftp_session.dart:
```diff
@@ -77,10 +110,10 @@ class FtpSession {
+    var address = (await _getIpAddress()).replaceAll('.', ',');
     sendResponse('227 Entering Passive Mode ($address,$p1,$p2)');
     dataListener!.first.then((socket) {
       dataSocket = socket;
```
i think having the `sendResponse('150 Opening data connection');` here is the better apporach since all commands that needs it looks like it wants that response.

abdelaziz-mahdy replied:
i meant after the `dataSocket = socket;` line

abdelaziz-mahdy commented on lib/ftp_session.dart:
```diff
@@ -77,10 +110,10 @@ class FtpSession {
     // I'm not sure what happened here, the client shows nothing if I comment out this line.
-    await Future.delayed(const Duration(microseconds: 100));
+    // await Future.delayed(const Duration(microseconds: 0));
+    var address = (await _getIpAddress()).replaceAll('.', ',');
```
what is the difference that the _getIpAddress make? since `String address = controlSocket.address.address.replaceAll('.', ',');` is alot cleaner.

lawnvi replied:
I not sure, I test 3:
```
1. String address = controlSocket.address.address.replaceAll('.', ',');

2. await Future.delayed(const Duration(microseconds: 100));
   String address = controlSocket.remoteAddress.address.replaceAll('.', ',');

3. await Future.delayed(const Duration(microseconds: 100));
   var address = (await _getIpAddress()).replaceAll('.', ',');
```
2 and 3 seem to respond with the same data.: `227 Entering Passive Mode (192,168,127,100,223,209)`, but linux show `Passive mode address mismatch.`

linux terminal context:
```
# _getIpAddress
ftp> ls
227 Entering Passive Mode (192,168,127,100,225,105)
150 Opening data connection
-rw-r--r-- 1 ftp ftp 8196 Jun 02 04:52 .DS_Store
drwxr-xr-x 1 ftp ftp 96 Jun 02 00:44 sefes12
drwxr-xr-x 1 ftp ftp 96 Jun 02 01:14 -a
-rw-r--r-- 1 ftp ftp 12410884 Jan 12 00:45 whisper-release.zip
-rw-r--r-- 1 ftp ftp 7787296 Jan 12 00:50 whisper-release.tar.xz
drwxr-xr-x 1 ftp ftp 96 Jun 02 00:44 未命名文件夹
-rw-r--r-- 1 ftp ftp 25182269 Jan 12 00:33 whisper-release.apk
drwxr-xr-x 1 ftp ftp 96 Jan 12 00:44 whisper.app
226 Transfer complete

# address or remoteAddress
ftp> ls
227 Entering Passive Mode (192,168,127,107,225,138)
Passive mode address mismatch.
ftp> ls
227 Entering Passive Mode (0,0,0,0,226,168)
Passive mode address mismatch.
ftp>
```

abdelaziz-mahdy replied:
Since you are testing using ftp command line , I will add the test cases to GitHub ci, 

The reason I didn't add them is that store file doesn't seem to work using the command line ftp on macos, it's just hangs 

If you tried it let me know if I implemented it wrongly on the test case

abdelaziz-mahdy commented:
i added the test cases and its cli, if all pass on main branch i will check if they pass on this branch too, if all good then i dont see a problem

abdelaziz-mahdy committed:
Merge branch 'main' of https://github.com/abdelaziz-mahdy/ftp_server into pr/lawnvi/3

abdelaziz-mahdy committed:
fix format

abdelaziz-mahdy committed:
Merge branch 'main' of https://github.com/abdelaziz-mahdy/ftp_server into pr/lawnvi/3

abdelaziz-mahdy committed:
format

abdelaziz-mahdy committed:
Merge branch 'main' of https://github.com/abdelaziz-mahdy/ftp_server into pr/lawnvi/3

abdelaziz-mahdy commented:
hello i updated your pr to my newest code i think no awaits for 100 mili is required anymore, but if you can test it that will be awesome, after you confirm i will merge and release the new version.

some of the test cases fails still on linux and windows, but i cant figure out why

abdelaziz-mahdy committed:
Merge branch 'main' of https://github.com/abdelaziz-mahdy/ftp_server into pr/lawnvi/3

abdelaziz-mahdy committed:
remove open data connection

abdelaziz-mahdy commented:
linux fails since it doesnt capture the return of the ftp server

lawnvi commented:
> hello i updated your pr to my newest code i think no awaits for 100 mili is required anymore, but if you can test it that will be awesome, after you confirm i will merge and release the new version.
> 
> some of the test cases fails still on linux and windows, but i cant figure out why

I'll go and give it a try.

abdelaziz-mahdy commented:
its already released 0.0.5 but you can test to make sure i didnt break anything 
Also, i am very interested in your whisper project, since the main objective of my package was to transfer files between my apps across platforms, so if you are able to package the auto-discovery and transfer, that will be awesome, since i think other people may like/need it too

lawnvi commented:
> its already released 0.0.5 but you can test to make sure i didnt break anything Also, i am very interested in your whisper project, since the main objective of my package was to transfer files between my apps across platforms, so if you are able to package the auto-discovery and transfer, that will be awesome, since i think other people may like/need it too

whisper discovers devices using [Bonsoir](https://pub-web.flutter-io.cn/packages/bonsoir).

I’m testing on Linux. It looks like something is wrong, but I’ll try to figure it out.

abdelaziz-mahdy commented:
> > its already released 0.0.5 but you can test to make sure i didnt break anything Also, i am very interested in your whisper project, since the main objective of my package was to transfer files between my apps across platforms, so if you are able to package the auto-discovery and transfer, that will be awesome, since i think other people may like/need it too
> 
> whisper discovers devices using [Bonsoir](https://pub-web.flutter-io.cn/packages/bonsoir).
> 
> I’m testing on Linux. It looks like something is wrong, but I’ll try to figure it out.

Let me know what is the problem, and feel free to open a pr for it 

I switched the placement of the data connection modes to only send that the mode is active after the data socket is acquired so this should fix the await problem 

And since macos is working and tests are good to go this means it may be something related to Linux, my only problem is that I can't test on it 

So if you are able to fix the test cases on Linux or windows that will be awesome ❤️


Also thank you for the discovery package

lawnvi commented:
> > > its already released 0.0.5 but you can test to make sure i didnt break anything Also, i am very interested in your whisper project, since the main objective of my package was to transfer files between my apps across platforms, so if you are able to package the auto-discovery and transfer, that will be awesome, since i think other people may like/need it too
> > 
> > 
> > whisper discovers devices using [Bonsoir](https://pub-web.flutter-io.cn/packages/bonsoir).
> > I’m testing on Linux. It looks like something is wrong, but I’ll try to figure it out.
> 
> Let me know what is the problem, and feel free to open a pr for it
> 
> I switched the placement of the data connection modes to only send that the mode is active after the data socket is acquired so this should fix the await problem
> 
> And since macos is working and tests are good to go this means it may be something related to Linux, my only problem is that I can't test on it
> 
> So if you are able to fix the test cases on Linux or windows that will be awesome ❤️
> 
> Also thank you for the discovery package

When Linux and Windows act as clients, sending the PASV and EPSV commands will cause blocking. Strangely, MacOS does not. This might be due to differences in client implementations, so I moved responses
`sendResponse('227 Entering Passive Mode ($address,$p1,$p2)');` 
back to the outer layer.

The response format tested on Linux is slightly different and has been modified. As for Windows, I haven't been able to find a way to use streams to get console output, but I discovered an alternative approach. You can see it in my new commit. I hope it gives you some inspiration.

abdelaziz-mahdy commented:
> > > > its already released 0.0.5 but you can test to make sure i didnt break anything Also, i am very interested in your whisper project, since the main objective of my package was to transfer files between my apps across platforms, so if you are able to package the auto-discovery and transfer, that will be awesome, since i think other people may like/need it too
> > > 
> > > 
> > > whisper discovers devices using [Bonsoir](https://pub-web.flutter-io.cn/packages/bonsoir).
> > > I’m testing on Linux. It looks like something is wrong, but I’ll try to figure it out.
> > 
> > 
> > Let me know what is the problem, and feel free to open a pr for it
> > I switched the placement of the data connection modes to only send that the mode is active after the data socket is acquired so this should fix the await problem
> > And since macos is working and tests are good to go this means it may be something related to Linux, my only problem is that I can't test on it
> > So if you are able to fix the test cases on Linux or windows that will be awesome ❤️
> > Also thank you for the discovery package
> 
> When Linux and Windows act as clients, sending the PASV and EPSV commands will cause blocking. Strangely, MacOS does not. This might be due to differences in client implementations, so I moved responses `sendResponse('227 Entering Passive Mode ($address,$p1,$p2)');` back to the outer layer.
> 
> The response format tested on Linux is slightly different and has been modified. As for Windows, I haven't been able to find a way to use streams to get console output, but I discovered an alternative approach. You can see it in my new commit. I hope it gives you some inspiration.

can you open a new pr? 

my problem is that the test cases fails to be ran on linux and windows

lawnvi commented:
Okay, I will tidy up the code after work and submit the PR.
